cmd/deploy/input_test.go

```diff
@@ -2,9 +2,279 @@ package main
+				},
+				"sha256:0cb474919526d040392883b84e5babb65a149cc605b89b117781ab94e88a5e86": {
+					Host:       "docker.io",
+					Repository: "foo/bar",
+					Tag:        "baz",
+					Digest:     "sha256:0cb474919526d040392883b84e5babb65a149cc605b89b117781ab94e88a5e86",
+				},
+				"sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e": {
+					Host:       "localhost:5000",
+					Repository: "example",
+					Tag:        "bar",
+					Digest:     "sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+				},
+			},
+			wantDigest: "nil",
+			wantErr:    "",
+		},
+		{
+			name: "basic fallback",
+			rawLookup: map[string]string{
+				"": "example",
+			},
+			wantLookup: map[ociregistry.Digest]registry.Reference{
+				"": {
+					Host:       "docker.io",
+					Repository: "library/example",
+				},
+			},
+			wantDigest: "nil",
+			wantErr:    "",
+		},
+		{
+			name: "digest fallback",
+			rawLookup: map[string]string{
+				"": "example@sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+			},
+			wantLookup: map[ociregistry.Digest]registry.Reference{
+				"": {
+					Host:       "docker.io",
+					Repository: "library/example",
+					Digest:     "sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+				},
+			},
+			wantDigest: "",
+			wantErr:    "",
+		},
+		{
+			name: "tag fallback",
+			rawLookup: map[string]string{
+				"": "example:tag",
+			},
+			wantLookup: map[ociregistry.Digest]registry.Reference{
+				"": {
+					Host:       "docker.io",
+					Repository: "library/example",
+					Tag:        "tag",
+				},
+			},
+			wantDigest: "",
+			wantErr:    "",
+		},
+		{
+			name: "bad digest",
+			rawLookup: map[string]string{
+				"invalid digest": "example:foo",
+			},
+			wantErr: `lookup key "invalid digest" invalid:`,
+		},
+		{
+			name: "bad reference",
+			rawLookup: map[string]string{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": "://#",
+			},
+			wantErr: "failed to parse lookup ref",
+		},
+		{
+			name: "digest mistmatch",
+			rawLookup: map[string]string{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": "example:foo@sha256:00f42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+			},
+			wantErr: "digest on lookup ref should either be omitted or match key:",
+		},
+	} {
+		t.Run(x.name, func(t *testing.T) {
+			lookups, lookupDigest, err := normalizeInputLookup(x.rawLookup)
+
+			if x.wantErr != "" {
+				if err == nil {
+					t.Fatalf("Expected error not returned: %s", x.wantErr)
+				}
+
+				if !strings.Contains(err.Error(), x.wantErr) {
+					t.Fatalf("Expected error doesn't match.\ngot:\n%q,\n\nexpected to contain:\n%q", err, x.wantErr)
+				}
+				return
+			}
+
+			// This allows consistent display of <nil>
+			var wantDigest *digest.Digest
+			if x.wantDigest != "nil" {
+				v := digest.Digest(x.wantDigest)
+				wantDigest = &v
+			}
```

Comment:
Is this purely cosmetic, or was there a functional reason for doing all this pointer juggling?  (in general, we want to avoid pointers where reasonable, and this seems like a place where we don't actually need the pointer except for _slightly_ better display of the error case, which IMO isn't something we should optimize for)

Reply:
Yeah, it's to avoid `nil` vs `<nil>` display

Reply:
Maybe it's just me but it would throw me off to have different representation of nil

Reply:
With `%q` it's a little stranger (since we'll get `"nil"`), but IMO that's still plenty clear especially if you check the test code to see exactly which conditional failed, especially since `nil` is an otherwise invalid digest.

Reply:
In other words, this will exhibit as either `got "something" expected "nil"` or `got "<nil>" expected "something"` and never `got "<nil>" expected "nil"` so I don't think it's ambiguous in the context of any single error returned from the test (since this one error will never give us _both_ `nil` representations at once).

Reply:
Done